contrib/cirrus/setup_environment.sh

```diff
@@ -6,16 +6,28 @@ source $(dirname $0)/lib.sh
 
 record_timestamp "env. setup start"
 
-req_env_var "
-USER $USER
-HOME $HOME
-ENVLIB $ENVLIB
-SCRIPT_BASE $SCRIPT_BASE
-CIRRUS_BUILD_ID $CIRRUS_BUILD_ID"
+req_env_var $(basename $0) "
+    USER $USER
+    HOME $HOME
+    GOSRC $GOSRC
+    SCRIPT_BASE $SCRIPT_BASE
+    SETUP_MARKER_FILEPATH $SETUP_MARKER_FILEPATH
+"
 
-[[ "$SHELL" =~ "bash" ]] || chsh -s /bin/bash
+# Ensure this script only executes successfully once and always logs ending timestamp
+[[ ! -e "$SETUP_MARKER_FILEPATH" ]] || exit 0
+exithandler() {
+    RET=$?
+    set +e
+    show_env_vars
+    record_timestamp "env. setup end"
+    echo "$(basename $0) exit status: $RET"
+    [[ "$RET" -eq "0" ]] && date +%s >> "SETUP_MARKER_FILEPATH"
+}
+trap exithandler EXIT
 
-cd "$CIRRUS_WORKING_DIR"  # for clarity of initial conditions
+# Must be bash, always bash
+[[ "$SHELL" =~ "bash" ]] || chsh -s /bin/bash
```

Comment:
This is almost certainly wrong. If it isn't, it's at least ultra-confusing. `chsh` changes the `/etc/passwd` entry, doing nothing at all to the currently-running shell, and not taking effect until subsequent logins. If that is your intended goal, I would like to see a comment saying so explicitly (so future maintainers don't have a WTF moment like I just did) and with some acknowledgment that this is done so that some-future-process will inherit this setting; please also include what possible circumstances you foresee in which this could happen, and an explanation of why it's so important for login shell to be bash.

Or, better yet, just remove these two lines.

Reply:
This was required early on with Cirrus-CI, it can possibly be removed.  IIRC, there wasn't a `CIRRUS_SHELL` variable and Ubuntu defaulted to `sh` or some-such was messing with executing individual commands from '.cirrus.yml' (as opposed to calling scripts).  I'll investigate and see if we can just strip it out.